garak/generators/nim.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python3
+
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""NVIDIA Inference Microservice LLM interface"""
+
+import json
+import logging
+import os
+import requests
+from typing import Union
+
+import backoff
+
+from garak import _config
+from garak.generators.base import Generator
+
+
+class NVHostedNimGenerator(Generator):
+    """Wrapper for NVIDIA-hosted NIMs. Expects NIM_API_KEY environment variable."""
+
+    supports_multiple_generations = False
+    generator_family_name = "NIM"
+    temperature = 0.1
+    top_p = 0.7
+
+    url = "https://integrate.api.nvidia.com/v1/chat/completions"
+
+    timeout = 60
+
+    def __init__(self, name=None, generations=10):
+        self.name = name
+        self.fullname = f"NIM {self.name}"
+        self.seed = _config.run.seed
+
+        if self.name is None:
+            raise ValueError("Please specify a NIM in model name (-n)")
+
+        super().__init__(name, generations=generations)
+
+        self.api_key = os.getenv("NIM_API_KEY", default=None)
+        if self.api_key is None:
+            raise ValueError(
+                'Put the NIM API key in the NIM_API_KEY environment variable (this was empty)\n \
+                e.g.: export NIM_API_KEY="nvapi-xXxXxXxXxXxXxXxXxXxX"'
+            )
```

Comment:
Should we go with [`NGC_API_KEY`](https://github.com/leondz/garak/blob/00c21b01b239c4a2174c81541619c0bfd6bac981/garak/generators/nemo.py#L41) as done with `nemo.py`?

`nvcf.py` uses a specifics [`NVCF_API_KEY`](https://github.com/leondz/garak/blob/00c21b01b239c4a2174c81541619c0bfd6bac981/garak/generators/nvcf.py#L45), not sure what the best pattern is here, I am also seeing some other non-nvidia tooling referring to [`NVIDIA_API_KEY`](https://haystack.deepset.ai/blog/haystack-nvidia-integration) and [NVIDIA docs](https://docs.nvidia.com/ai-enterprise/nim-llm/latest/deploy-helm.html) suggesting `NGC_CLI_API_KEY` when using helm for NIM deployments so I guess there is not really a standard per se.

Reply:
From what I've seen NGC CLI API keys are not the same as NVIDIA-hosted NIM API keys. There are a few moving parts in the ecosystem.
* NVCF is a different service & API with a two-key auth.
* NGC is more like a CSP, with tooling and APIs around it. 
* NIM is an inference container format. 
 
NIMs can run on NGC, but can also run locally, or even on NVCF (if they fit on one node). 

This PR provides a class for NIMs hosted by NVIDIA (which afaik run on the NGC clusters). Generator classes for locally-hosted NIMs etc. can also go in this module - tagging [this API reference page](https://docs.nvidia.com/ai-enterprise/nim-llm/latest/quickstart/mixtral-8x7b-instruct.html#launch-microservice) for later

Reply:
FWIW, my `NGC_API_KEY` did not work, but I was able to get the NIM API working with my `NVCF_API_KEY`.

Reply:
Nice. They do look similar. I wish I knew if all NIM calls could be done using NVCF keys. I might just leave this in here for now pending a shift to conf-based key specification with per-generator `api_key`, obviating cramming everything into env var space -- see #643.